File: lua/codetyper/transport/protocol.lua
---@mod codetyper.transport.protocol JSON-RPC protocol for agent communication
---@brief [[
--- Defines the JSON-RPC protocol used for Lua <-> Python agent communication.
--- This module handles serialization, deserialization, and protocol validation.
---
--- Protocol Format:
---   Request:  {"jsonrpc": "2.0", "method": str, "params": dict, "id": int}
---   Response: {"jsonrpc": "2.0", "result": any, "id": int}
---   Error:    {"jsonrpc": "2.0", "error": {"code": int, "message": str}, "id": int}
---@brief ]]

local M = {}

---@class RPCRequest
---@field jsonrpc string Always "2.0"
---@field method string Method name
---@field params table Parameters
---@field id number|string Request ID

---@class RPCResponse
---@field jsonrpc string Always "2.0"
---@field result? any Result on success
---@field error? RPCError Error on failure
---@field id number|string|nil Request ID

---@class RPCError
---@field code number Error code
---@field message string Error message
---@field data? any Additional error data

-- Standard JSON-RPC error codes
M.ErrorCodes = {
  PARSE_ERROR = -32700,
  INVALID_REQUEST = -32600,
  METHOD_NOT_FOUND = -32601,
  INVALID_PARAMS = -32602,
  INTERNAL_ERROR = -32603,
  -- Custom server errors
  INTENT_CLASSIFICATION_FAILED = -32001,
  PLAN_CONSTRUCTION_FAILED = -32002,
  VALIDATION_FAILED = -32003,
  CONTEXT_ERROR = -32004,
  FORMATTING_ERROR = -32005,
}

-- Available methods
M.Methods = {
  CLASSIFY_INTENT = "classify_intent",
  BUILD_PLAN = "build_plan",
  VALIDATE_PLAN = "validate_plan",
  FORMAT_OUTPUT = "format_output",
  PING = "ping",
}

-- Intent types (must match Python IntentType enum)
M.IntentType = {
  ASK = "ask",
  CODE = "code",
  REFACTOR = "refactor",
  DOCUMENT = "document",
  FIX = "fix",
  EXPLAIN = "explain",
  TEST = "test",
  UNKNOWN = "unknown",
}

-- Action types (must match Python ActionType enum)
M.ActionType = {
  READ = "read",
  WRITE = "write",
  EDIT = "edit",
  DELETE = "delete",
  RENAME = "rename",
  CREATE_DIR = "create_dir",
}

-- Request ID counter
local request_id = 0

---Generate a new request ID
---@return number
local function next_id()
  request_id = request_id + 1
  return request_id
end

---Create a JSON-RPC request
---@param method string Method name
---@param params table Parameters
---@return RPCRequest
function M.make_request(method, params)
  return {
    jsonrpc = "2.0",
    method = method,
    params = params or {},
    id = next_id(),
  }
end

---Serialize a request to JSON string
---@param request RPCRequest
---@return string
function M.serialize_request(request)
  local ok, json = pcall(vim.json.encode, request)
  if not ok then
    error("Failed to serialize request: " .. tostring(json))
  end
  return json
end

---Deserialize a response from JSON string
---@param raw string
---@return RPCResponse|nil response
---@return string|nil error
function M.deserialize_response(raw)
  local ok, response = pcall(vim.json.decode, raw)
  if not ok then
    return nil, "Failed to parse response: " .. tostring(response)
  end

  -- Validate response structure
  if type(response) ~= "table" then
    return nil, "Response is not an object"
  end

  if response.jsonrpc ~= "2.0" then
    return nil, "Invalid JSON-RPC version"
  end

  return response, nil
end

---Check if a response is an error
---@param response RPCResponse
---@return boolean
function M.is_error(response)
  return response.error ~= nil
end

---Extract error message from response
---@param response RPCResponse
---@return string
function M.get_error_message(response)
  if not response.error then
    return "Unknown error"
  end
  local msg = response.error.message or "Unknown error"
  if response.error.code then
    msg = string.format("[%d] %s", response.error.code, msg)
  end
  return msg
end

---Build IntentRequest params
---@param context string Buffer content and surrounding context
---@param prompt string User's prompt
---@param files? string[] Referenced file paths
---@return table
function M.build_intent_request(context, prompt, files)
  return {
    context = context,
    prompt = prompt,
    files = files or {},
  }
end

---Build PlanRequest params
---@param intent string Intent type
---@param context string Context
---@param files table<string, string> File path -> content map
---@return table
function M.build_plan_request(intent, context, files)
  return {
    intent = intent,
    context = context,
    files = files,
  }
end

---Build ValidationRequest params
---@param plan table Plan response from build_plan
---@param original_files table<string, string> Original file contents
---@return table
function M.build_validation_request(plan, original_files)
  return {
    plan = plan,
    original_files = original_files,
  }
end

---Build FormatRequest params
---@param format_type string "plan"|"diff"|"error"
---@param data table Data to format
---@return table
function M.build_format_request(format_type, data)
  return {
    type = format_type,
    data = data,
  }
end

---Parse IntentResponse from result
---@param result table Raw result from response
---@return table IntentResponse
function M.parse_intent_response(result)
  return {
    intent = result.intent,
    confidence = result.confidence,
    reasoning = result.reasoning,
    needs_clarification = result.needs_clarification or false,
    clarification_questions = result.clarification_questions or {},
  }
end

---Parse PlanResponse from result
---@param result table Raw result from response
---@return table PlanResponse
function M.parse_plan_response(result)
  local steps = {}
  for _, step in ipairs(result.steps or {}) do
    table.insert(steps, {
      id = step.id,
      action = step.action,
      target = step.target,
      params = step.params or {},
      depends_on = step.depends_on or {},
    })
  end

  local rollback_steps = {}
  for _, step in ipairs(result.rollback_steps or {}) do
    table.insert(rollback_steps, {
      id = step.id,
      action = step.action,
      target = step.target,
      params = step.params or {},
      depends_on = step.depends_on or {},
    })
  end

  return {
    steps = steps,
    needs_clarification = result.needs_clarification or false,
    clarification_questions = result.clarification_questions or {},
    rollback_steps = rollback_steps,
  }
end

---Parse ValidationResponse from result
---@param result table Raw result from response
---@return table ValidationResponse
function M.parse_validation_response(result)
  return {
    valid = result.valid,
    errors = result.errors or {},
    warnings = result.warnings or {},
  }
end

---Validate IntentRequest params
---@param params table
---@return boolean valid
---@return string|nil error
function M.validate_intent_request(params)
  if type(params.context) ~= "string" then
    return false, "Missing or invalid 'context' field"
  end
  if type(params.prompt) ~= "string" then
    return false, "Missing or invalid 'prompt' field"
  end
  if params.files and type(params.files) ~= "table" then
    return false, "'files' must be a list"
  end
  return true, nil
end

---Validate PlanRequest params
---@param params table
---@return boolean valid
---@return string|nil error
function M.validate_plan_request(params)
  if type(params.intent) ~= "string" then
    return false, "Missing or invalid 'intent' field"
  end
  if type(params.context) ~= "string" then
    return false, "Missing or invalid 'context' field"
  end
  if type(params.files) ~= "table" then
    return false, "Missing or invalid 'files' field"
  end
  return true, nil
end

return M

File: agent/protocol.py
"""
JSON-RPC protocol implementation for Lua <-> Agent communication.

This module defines:
- Request/response message formats
- Method routing and dispatch
- Error handling with structured error codes

Protocol:
    Request:  {"jsonrpc": "2.0", "method": str, "params": dict, "id": int}
    Response: {"jsonrpc": "2.0", "result": any, "id": int}
    Error:    {"jsonrpc": "2.0", "error": {"code": int, "message": str}, "id": int}

Methods:
    - classify_intent: Classify user intent
    - build_plan: Build execution plan
    - validate_plan: Validate plan
    - format_output: Format output for display

Error Codes:
    - -32700: Parse error
    - -32600: Invalid request
    - -32601: Method not found
    - -32602: Invalid params
    - -32603: Internal error
    - -32000 to -32099: Server errors (custom)
"""

from typing import Any, Callable, Dict, Optional, TypeVar, Union
from dataclasses import dataclass, asdict
import json
import traceback

from .schemas import (
    IntentRequest,
    IntentResponse,
    PlanRequest,
    PlanResponse,
    ValidationRequest,
    ValidationResponse,
    serialize,
    deserialize,
)


# Type alias for method handlers
MethodHandler = Callable[[Dict[str, Any]], Any]

# Registry of method handlers
_method_registry: Dict[str, MethodHandler] = {}


class RPCErrorCodes:
    """Standard JSON-RPC error codes."""
    PARSE_ERROR = -32700
    INVALID_REQUEST = -32600
    METHOD_NOT_FOUND = -32601
    INVALID_PARAMS = -32602
    INTERNAL_ERROR = -32603
    # Custom server errors
    INTENT_CLASSIFICATION_FAILED = -32001
    PLAN_CONSTRUCTION_FAILED = -32002
    VALIDATION_FAILED = -32003
    CONTEXT_ERROR = -32004
    FORMATTING_ERROR = -32005


@dataclass
class RPCError(Exception):
    """JSON-RPC error with code and message."""
    code: int
    message: str
    data: Optional[Any] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to JSON-RPC error object."""
        error = {
            "code": self.code,
            "message": self.message,
        }
        if self.data is not None:
            error["data"] = self.data
        return error


@dataclass
class RPCRequest:
    """Parsed JSON-RPC request."""
    jsonrpc: str
    method: str
    params: Dict[str, Any]
    id: Union[int, str, None]

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "RPCRequest":
        """Parse request from dict."""
        return cls(
            jsonrpc=data.get("jsonrpc", ""),
            method=data.get("method", ""),
            params=data.get("params", {}),
            id=data.get("id"),
        )

    def validate(self) -> None:
        """Validate request structure."""
        if self.jsonrpc != "2.0":
            raise RPCError(
                RPCErrorCodes.INVALID_REQUEST,
                "Invalid JSON-RPC version, expected '2.0'",
            )
        if not self.method:
            raise RPCError(
                RPCErrorCodes.INVALID_REQUEST,
                "Missing 'method' field",
            )
        if not isinstance(self.method, str):
            raise RPCError(
                RPCErrorCodes.INVALID_REQUEST,
                "'method' must be a string",
            )
        if self.params is not None and not isinstance(self.params, dict):
            raise RPCError(
                RPCErrorCodes.INVALID_PARAMS,
                "'params' must be an object",
            )


def make_response(result: Any, request_id: Union[int, str, None]) -> Dict[str, Any]:
    """Create a JSON-RPC success response."""
    return {
        "jsonrpc": "2.0",
        "result": serialize(result) if hasattr(result, "__dataclass_fields__") else result,
        "id": request_id,
    }


def make_error_response(
    error: RPCError,
    request_id: Union[int, str, None] = None,
) -> Dict[str, Any]:
    """Create a JSON-RPC error response."""
    return {
        "jsonrpc": "2.0",
        "error": error.to_dict(),
        "id": request_id,
    }


def register_method(name: str, handler: MethodHandler) -> None:
    """
    Register a method handler.

    Args:
        name: Method name (e.g., "classify_intent")
        handler: Function that takes params dict and returns result
    """
    _method_registry[name] = handler


def unregister_method(name: str) -> None:
    """Unregister a method handler."""
    _method_registry.pop(name, None)


def get_registered_methods() -> list:
    """Get list of registered method names."""
    return list(_method_registry.keys())


def handle_request(request: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle a JSON-RPC request and return a response.

    Args:
        request: Parsed JSON-RPC request dict

    Returns:
        JSON-RPC response dict
    """
    request_id = request.get("id")

    try:
        # Parse and validate request
        rpc_request = RPCRequest.from_dict(request)
        rpc_request.validate()

        # Look up method handler
        handler = _method_registry.get(rpc_request.method)
        if handler is None:
            raise RPCError(
                RPCErrorCodes.METHOD_NOT_FOUND,
                f"Method '{rpc_request.method}' not found",
            )

        # Execute handler
        result = handler(rpc_request.params)

        return make_response(result, request_id)

    except RPCError as e:
        return make_error_response(e, request_id)
    except Exception as e:
        # Wrap unexpected errors
        return make_error_response(
            RPCError(
                RPCErrorCodes.INTERNAL_ERROR,
                str(e),
                data=traceback.format_exc(),
            ),
            request_id,
        )


def parse_request(raw: str) -> Dict[str, Any]:
    """
    Parse a raw JSON string into a request dict.

    Args:
        raw: Raw JSON string

    Returns:
        Parsed request dict

    Raises:
        RPCError: If JSON parsing fails
    """
    try:
        return json.loads(raw)
    except json.JSONDecodeError as e:
        raise RPCError(
            RPCErrorCodes.PARSE_ERROR,
            f"Invalid JSON: {e.msg}",
        )


def format_response(response: Dict[str, Any]) -> str:
    """
    Format a response dict as JSON string.

    Args:
        response: Response dict

    Returns:
        JSON string
    """
    return json.dumps(response, ensure_ascii=False)


# ============================================================
# Method Handlers
# ============================================================

def _handle_classify_intent(params: Dict[str, Any]) -> Dict[str, Any]:
    """Handle classify_intent method."""
    from .intent import IntentClassifier
    from .schemas import IntentRequest

    try:
        request = deserialize(IntentRequest, params)
    except (KeyError, TypeError) as e:
        raise RPCError(
            RPCErrorCodes.INVALID_PARAMS,
            f"Invalid params for classify_intent: {e}",
        )

    classifier = IntentClassifier()
    response = classifier.classify(request)
    return serialize(response)


def _handle_build_plan(params: Dict[str, Any]) -> Dict[str, Any]:
    """Handle build_plan method."""
    from .planner import Planner
    from .schemas import PlanRequest

    try:
        request = deserialize(PlanRequest, params)
    except (KeyError, TypeError) as e:
        raise RPCError(
            RPCErrorCodes.INVALID_PARAMS,
            f"Invalid params for build_plan: {e}",
        )

    planner = Planner()
    response = planner.build_plan(request)
    return serialize(response)


def _handle_validate_plan(params: Dict[str, Any]) -> Dict[str, Any]:
    """Handle validate_plan method."""
    from .validator import PlanValidator
    from .schemas import ValidationRequest

    try:
        request = deserialize(ValidationRequest, params)
    except (KeyError, TypeError) as e:
        raise RPCError(
            RPCErrorCodes.INVALID_PARAMS,
            f"Invalid params for validate_plan: {e}",
        )

    validator = PlanValidator()
    response = validator.validate(request)
    return serialize(response)


def _handle_format_output(params: Dict[str, Any]) -> Dict[str, Any]:
    """Handle format_output method."""
    from .formatter import OutputFormatter

    formatter = OutputFormatter()
    format_type = params.get("type", "plan")
    data = params.get("data", {})

    if format_type == "plan":
        from .schemas import PlanResponse
        plan = deserialize(PlanResponse, data)
        return {"formatted": formatter.format_plan(plan)}
    elif format_type == "diff":
        return {"formatted": formatter.format_diff(
            data.get("original", ""),
            data.get("modified", ""),
        )}
    elif format_type == "error":
        return {"formatted": formatter.format_error(
            Exception(data.get("message", "Unknown error")),
            data.get("context"),
        )}
    else:
        raise RPCError(
            RPCErrorCodes.INVALID_PARAMS,
            f"Unknown format type: {format_type}",
        )


def _handle_ping(params: Dict[str, Any]) -> Dict[str, Any]:
    """Handle ping method for health checks."""
    return {"status": "ok", "version": "0.1.0"}


# ============================================================
# Memory Methods
# ============================================================

# Global memory state (initialized per project)
_memory_graph = None
_memory_storage = None
_memory_learners = None


def _ensure_memory_initialized(project_root: str):
    """Ensure memory system is initialized for the project."""
    global _memory_graph, _memory_storage, _memory_learners

    from .memory import MemoryGraph, MemoryStorage, PatternLearner, ConventionLearner, CorrectionLearner

    if _memory_storage is None or str(_memory_storage.project_root) != project_root:
        _memory_storage = MemoryStorage(project_root)
        _memory_graph = _memory_storage.load()
        _memory_learners = {
            "pattern": PatternLearner(_memory_graph),
            "convention": ConventionLearner(_memory_graph),
            "correction": CorrectionLearner(_memory_graph),
        }


def _handle_memory_learn(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle memory_learn method.

    Learns from an event (edit, correction, approval, etc.)

    Params:
        project_root: str - Project root directory
        event_type: str - Type of event (edit, correction, approval, rejection)
        data: dict - Event data
    """
    project_root = params.get("project_root", ".")
    event_type = params.get("event_type", "")
    event_data = params.get("data", {})

    if not event_type:
        raise RPCError(RPCErrorCodes.INVALID_PARAMS, "event_type is required")

    _ensure_memory_initialized(project_root)

    from .memory.learners import Event

    event = Event(type=event_type, data=event_data)

    # Let all learners observe the event
    for learner in _memory_learners.values():
        learner.observe(event)

    # Save changes
    saved = _memory_storage.save(_memory_graph)

    return {
        "learned": True,
        "saved": saved,
        "node_count": len(_memory_graph._nodes),
    }


def _handle_memory_query(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle memory_query method.

    Query the memory graph for relevant knowledge.

    Params:
        project_root: str - Project root directory
        node_type: str|None - Filter by node type
        content_pattern: str|None - Filter by content pattern
        limit: int - Maximum results (default 20)
    """
    project_root = params.get("project_root", ".")
    node_type_str = params.get("node_type")
    content_pattern = params.get("content_pattern")
    limit = params.get("limit", 20)

    _ensure_memory_initialized(project_root)

    from .memory.graph import NodeType

    node_type = None
    if node_type_str:
        try:
            node_type = NodeType(node_type_str)
        except ValueError:
            pass  # Invalid type, ignore filter

    nodes = _memory_graph.query(node_type=node_type, content_pattern=content_pattern)

    # Limit results
    nodes = nodes[:limit]

    return {
        "nodes": [
            {
                "id": n.id,
                "type": n.type.value,
                "content": n.content,
                "metadata": n.metadata,
            }
            for n in nodes
        ],
        "total": len(nodes),
    }


def _handle_memory_get_context(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle memory_get_context method.

    Get formatted memory context for LLM prompts.

    Params:
        project_root: str - Project root directory
        context_type: str - Type of context (patterns, conventions, corrections, all)
        max_tokens: int - Maximum tokens budget (default 2000)
    """
    project_root = params.get("project_root", ".")
    context_type = params.get("context_type", "all")
    max_tokens = params.get("max_tokens", 2000)

    _ensure_memory_initialized(project_root)

    from .memory.graph import NodeType

    lines = []
    char_budget = max_tokens * 4  # Rough estimate: 4 chars per token

    # Gather relevant nodes based on context type
    if context_type in ("patterns", "all"):
        patterns = _memory_graph.query(node_type=NodeType.PATTERN)
        if patterns:
            lines.append("## Learned Patterns")
            for p in patterns[:10]:
                lines.append(f"- {p.content}")

    if context_type in ("conventions", "all"):
        conventions = _memory_graph.query(node_type=NodeType.CONVENTION)
        if conventions:
            lines.append("\n## Project Conventions")
            for c in conventions[:10]:
                lines.append(f"- {c.content}")

    if context_type in ("corrections", "all"):
        corrections = _memory_graph.query(node_type=NodeType.CORRECTION)
        if corrections:
            lines.append("\n## Past Corrections (avoid these mistakes)")
            for c in corrections[:5]:
                lines.append(f"- {c.content}")

    context_text = "\n".join(lines)

    # Truncate if needed
    if len(context_text) > char_budget:
        context_text = context_text[:char_budget] + "\n... (truncated)"

    return {
        "context": context_text,
        "char_count": len(context_text),
        "estimated_tokens": len(context_text) // 4,
    }


def _handle_memory_stats(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle memory_stats method.

    Get statistics about the memory system.

    Params:
        project_root: str - Project root directory
    """
    project_root = params.get("project_root", ".")

    _ensure_memory_initialized(project_root)

    from .memory.graph import NodeType

    # Count nodes by type
    type_counts = {}
    for node_type in NodeType:
        count = len(_memory_graph.query(node_type=node_type))
        if count > 0:
            type_counts[node_type.value] = count

    storage_info = _memory_storage.get_storage_info()

    return {
        "node_count": len(_memory_graph._nodes),
        "edge_count": len(_memory_graph._edges),
        "type_counts": type_counts,
        "storage": storage_info,
    }


def _handle_memory_clear(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle memory_clear method.

    Clear all memory for the project.

    Params:
        project_root: str - Project root directory
    """
    project_root = params.get("project_root", ".")

    _ensure_memory_initialized(project_root)

    _memory_graph.clear()
    deleted = _memory_storage.delete()

    return {
        "cleared": True,
        "file_deleted": deleted,
    }


# Register built-in methods
register_method("classify_intent", _handle_classify_intent)
register_method("build_plan", _handle_build_plan)
register_method("validate_plan", _handle_validate_plan)
register_method("format_output", _handle_format_output)
register_method("ping", _handle_ping)

# Register memory methods
register_method("memory_learn", _handle_memory_learn)
register_method("memory_query", _handle_memory_query)
register_method("memory_get_context", _handle_memory_get_context)
register_method("memory_stats", _handle_memory_stats)
register_method("memory_clear", _handle_memory_clear)
